ngramoffset.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2021 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zoekt
+
+import (
+	"log"
+	"reflect"
+	"sort"
+)
+
+// simple interface type check for implementations
+var (
+	_ ngramOffset = mappedNgramOffset{}
+	_ ngramOffset = &arrayNgramOffset{}
+)
+
+type ngramOffset interface {
+	Get(ngram) simpleSection
+	DumpMap() map[ngram]simpleSection
+	Len() int
+}
+
+func makeNgramOffsetEmpty() ngramOffset {
+	return mappedNgramOffset(map[ngram]simpleSection{})
+}
+
+func makeNgramOffset(ngrams map[ngram]simpleSection) ngramOffset {
+	if len(ngrams) > 0 {
+		a := makeArrayNgramOffset(ngrams)
+		m := a.DumpMap()
+		if !reflect.DeepEqual(m, ngrams) {
+			log.Printf("arrayNgramOffset creation failed!")
+		} else {
+			return a
+		}
+	}
+	return mappedNgramOffset(ngrams)
+}
+
+type mappedNgramOffset map[ngram]simpleSection
+
+func (m mappedNgramOffset) Get(gram ngram) simpleSection {
+	return m[gram]
+}
+
+func (m mappedNgramOffset) DumpMap() map[ngram]simpleSection {
+	return m
+}
+
+func (m mappedNgramOffset) Len() int {
+	return len(m)
+}
+
+// arrayNgramOffset uses split linear arrays and binary search for Get()
```

Comment:
Can you elaborate in this comment on why we're using split arrays of uint32s rather than a single array of uint64s?

Reply:
Done.